Anomaly Detection Model on Time Series data (#6137)
section-engineering-education/engineering-education: content/articles/anomaly-detection-model-on-time-series-data/index.md

- [Dataset splitting](#dataset-splitting)
- [Initializing the model](#initializing-the-model)
- [Model fitting](#model-fitting)
- [Making predictions using the test dataset](#making-predictions-using-the-test-dataset)
- [Plotting diagram](#plotting-diagram)
- [Understanding the predicted values](#understanding-the-predicted-values)
- [Detecting anomalies](#detecting-anomalies)
- [Scatter plot](#scatter-plot)
- [Conclusion](#conclusion)
- [References](#references)

### Prerequisites
To follow along with this tutorial, a reader should:

- Have some knowledge on [time series](/engineering-education/introduction-to-time-series/)
- Understand [time series decomposition in python](/engineering-education/time-series-decomposition-in-python/)
- Know how to build a [time series model](/engineering-education/building-a-time-series-weather-forecasting-application-in-python/)
- Use [Google Colab notebook](https://research.google.com/)

### Time series dataset
The dataset used in this tutorial is the New York Taxi dataset. The dataset was recorded on a half-hour basis for 6 months. The dataset shows the number of active taxi drivers in New York city every half-hour. 

> We will use this dataset to build a times series model using Facebook Prophet. The model will be used to predict the number of active taxi drivers in New York. 

Reviewer: ```suggestion
We will use this dataset to build a times series model using Facebook Prophet. We will use the model to predict the number of active taxi drivers in New York. 
```

Reviewer: Always use active voice

```python
df
```
The output is shown below:

![Resampled dataset](/engineering-education/anomaly-detection-model-on-time-series-data/resampled-dataset.jpg)

From the image above, we have resampled our dataset and the new value is 5160.

### Plotting a line chart
We will visualize this dataset using the line chart. We will plot this diagram using the `plotly.express` library.

```python
fig = px.line(df.reset_index(), x='timestamp', y='value', title='NYC Taxi Demand')
fig.update_xaxes(
 rangeslider_visible=True,
 rangeselector=dict(
 buttons=list([
 dict(count=1, label='1y', step='year', stepmode='backward'),
 dict(count=2, label='3y', step='year', stepmode='backward'),
 dict(count=3, label='5y', step='year', stepmode='backward'),
 dict(step='all')
        ])
    )
)
fig.show()
```
From the code above, We have initialized the x-axis as `timestamp`. The y-axis is the `value`. And the title of the line chart is `NYC Taxi Demand`. We have added labels that can show the data points within 1 year, 3 years, or 5 years. 

The diagram is shown below:

![Line chart](/engineering-education/anomaly-detection-model-on-time-series-data/line-chart.jpg)

> The image above shows the number of active taxis with a period of July 2014 to Jan 2015. 

Reviewer: ```suggestion
The image above shows the number of active taxis within the period of July 2014 to Jan 2015. 
```

```
Let's rename our dataset column.

```python
taxi_df = df.reset_index()[['timestamp', 'value']].rename({'timestamp':'ds', 
 'value':'y'}, 
 axis='columns')
```

In the code above, we have renamed our `timestamp` column to `ds` and `value` column to `y`. To check the structure of the renamed dataset, use this code:

```python
taxi_df.head()
```
The output is shown below:

![Renamed columns](/engineering-education/anomaly-detection-model-on-time-series-data/renamed-columns.jpg)

### Dataset splitting
We need to split our dataset into two sets. One set will be used to train our time series model and the other one will test the model.

```python
train_dataset = taxi_df[(taxi_df['ds']>='2014-07-01')&(taxi_df['ds']<='2015-01-27')]
test_dataset = taxi_df[(taxi_df['ds']>'2015-01-27')]
```
From the code above, the dataset between `2014-07-01` and `2015-01-27` will be used for training. The test dataset contains timestamp values past `2015-01-27`. The model will use the test dataset to make predictions.

Let's initialize our model.

### Initializing the model
We will initialize our model using the following code:

```python
model = Prophet(changepoint_range=0.95)
```

> We use a `changepoint_range=0.95` to increase the confidence interval of our output. The confidence interval determines if a prediction is right or wrong. The higher confidence interval increases the chances of the model making accurate predictions

Reviewer: ```suggestion
We use a `changepoint_range=0.95` to increase the confidence interval of our output. The confidence interval determines if a prediction is right or wrong. The higher confidence interval increases the chances of the model making accurate predictions.
```

To understand the confidence interval in detail, read this [article](https://towardsdatascience.com/confidence-intervals-for-time-series-data-sharpestminds-slack-chats-4b05e211943a)

Let's fit the model to our train dataset.

### Model fitting
We fit the model to the training dataset so that the model can learn from it. 

```python
model.fit(train_dataset)
```

> This will train the model and the model will be ready to make future predictions.

Reviewer: ```suggestion
This code will train the model. After training, the model will now be ready to make future predictions.
```

### Making predictions using the test dataset
The test dataset contains timestamp values past `2015-01-27`. The test dataset has 120 hours. We are forecasting for the 120 hours in the test dataset.

```python
future = model.make_future_dataframe(periods=120, freq='H')
future.tail()
```

> From the code above `periods=120` shows the 120 hours recorded in the test dataset. `freq='H'` will give the prediction values in hourly-interval. `future.tail()` will show the five last rows of the `ds` column. This is the last five hours the model will predict.

Reviewer: ```suggestion
From the code above, `periods=120` shows the 120 hours recorded in the test dataset. `freq='H'` will give the prediction values in hourly-interval. `future.tail()` will show the five last rows of the `ds` column. It shows the last five hours the model will predict.
```

outcome.head()
```
The new data frame is shown below:

![New dataframe](/engineering-education/anomaly-detection-model-on-time-series-data/new-dataframe.jpg)

We now use these columns to detect anomalies.

### Detecting anomalies
We will use the following logic to identify anomalies.

```python
outcome['anomaly'] = outcome.apply(lambda x: 'Yes' if(np.abs(x['error'])>1.5*x['uncertainity']) else 'No',axis=1)
```
From the code above, a predicted value `yhat` is an anomaly if the absolute error value (error) is greater than 1.5 multiplied by the uncertainty value. The `outcome.apply` method will apply the logic and detect the anomalies.

We can now create a scatter plot to show the detected anomalies.

### Scatter plot
To create a scatter plot, use code snippet:

```python
fig = px.scatter(outcome.reset_index(), x='ds', y='y', 
 color='anomaly',title='NYC Taxi Demand')
fig.update_xaxes(
 rangeslider_visible=True,
 rangeselector=dict(
 buttons=list([
 dict(count=1, label='1y', step='year', stepmode='backward'),
 dict(count=2, label='3y', step='year', stepmode='backward'),
 dict(count=3, label='5y', step='year', stepmode='backward'),
 dict(step='all')
        ])
    )
)
fig.show()
```

> From the code above we have set `ds` as the x-axis and `y` as the y-axis. The title of the scatter plot is `NYC Taxi Demand`. Also, the anomalies detected will be given a different color. When the code is executed, it output the following diagram:

Reviewer: ```suggestion
From the code above, we have set `ds` as the x-axis and `y` as the y-axis. The title of the scatter plot is `NYC Taxi Demand`. Also, the anomalies detected will be given a different color. When the code is executed, it outputs the following diagram:
```

![Scatter plot](/engineering-education/anomaly-detection-model-on-time-series-data/scatter-plot.jpg)

In the image above, the detected anomalies have a red color. The normal data points have a blue color. The scatter plot has helped us to detect the outliers available in our dataset.

### Conclusion

> In this tutorial, we have learned anomaly detection on time series data. We used the New York Taxi dataset to train our model. Using the Facebook Prophet library, we were able to build a time series model. The model was used to predict the number of active taxi drivers in New York.
> 
> We then used a scatter plot to detect anomalies in our dataset. Using this tutorial, a reader should be able to detect anomalies on time series data.

Reviewer: ```suggestion
In this tutorial, we have learned anomaly detection on time series data. We used the New York Taxi dataset to train our model. Using the Facebook Prophet library, we built a time series model. The model was used to predict the number of active taxi drivers in New York.

We then used a scatter plot to detect anomalies in our dataset. Using this tutorial, a reader should detect anomalies on time series data.
```
